### PR TITLE
LR Scheduler fixes

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Loading a model trained with `cosine` or `linear` LR schedules was throwing an exception on PyTorch 1.6. This is because pytorch 1.6 broke our method for forwarding an LR schedule with `step(num_epochs)`.

List of changes:
- Migrates us to the new sanctioned LR loading
- Works on both pytorch 1.4 and 1.6 (didn't bother testing 1.5.1)
- Adds a new `test_agents/unigram` model, which is a torch generator agent that always outputs fixed distributions. This is useful to have a TA that has extremely few parameters (tens or hundreds) and can be trained extremely fast. (Also added its own CI for this)
- Add end2end tests for select LR schedulers
- Add resume LR checks for select LR schedulers.

**Testing steps**
- Ran select unit tests in both 1.4 and 1.6.
- New CI

### CODE REVIEW

File: parlai/agents/test_agents/unigram.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+UnigramAgent always predictions the same vocabulary distribution.
```
Comment (klshuster) [inline parlai/agents/test_agents/unigram.py:8]: nit: `always predicts`

File: parlai/agents/test_agents/unigram.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
```
Comment (klshuster) [inline parlai/agents/test_agents/unigram.py:1]: nice example of how easy it is to subclass TGA

File: parlai/nn/lr_scheduler.py [hunk 3]
```diff
@@ -48,21 +48,32 @@ def __init__(self, hard_reset, warmup_updates, warmup_rate):
...
         else:
             self.warmup_scheduler = None
 
+    def get_last_lr(self):
+        s = self.warmup_scheduler if self._is_lr_warming_up() else self.scheduler
+        try:
+            # pytorch 1.5 or newer
+            return s.get_last_lr()[0]
+        except AttributeError:
+            # pytorch 1.4 or older
```
Comment (klshuster) [inline parlai/nn/lr_scheduler.py:74]: maybe add a TODO here to remove this check if we ever stop supporting 1.4?

File: parlai/nn/lr_scheduler.py [hunk 4]
```diff
@@ -87,12 +98,17 @@ def load_state(self, states):
         """
         Load state of scheduler from states.
         """
-        if self.scheduler and 'lr_scheduler' in states:
-            self.scheduler.load_state_dict(states['lr_scheduler'])
         if states.get('warmup_scheduler') and getattr(self, 'warmup_scheduler', None):
             self.warmup_scheduler.load_state_dict(states['warmup_scheduler'])
+        if self.scheduler and 'lr_scheduler' in states:
```
Comment (klshuster) [inline parlai/nn/lr_scheduler.py:103]: is there any reason for moving this?

File: parlai/nn/lr_scheduler.py [hunk 5]
```diff
@@ -480,13 +501,13 @@ def __init__(
     def _linear_lr(self, step):
         # this multiplicative factor ensures linear decay rate
         # lr_mult = float(self.max_lr_steps - step - 1) / float(self.max_lr_steps - step)
-        lr_mult = max(0.0, 1.0 - step / self.max_lr_steps)
+        lr_mult = max(0.0, 1e-6 + (1.0 - step / self.max_lr_steps) * (1 - 1e-6))
```
Comment (klshuster) [inline parlai/nn/lr_scheduler.py:504]: what does this do?

**Review by klshuster — APPROVED** [review]
thanks for adding a bunch of tests too

File: parlai/nn/lr_scheduler.py (hunk 4 above)
Comment (stephenroller) [reply parlai/nn/lr_scheduler.py:103]: Ya idk something crashed for me but I don't remember.

File: parlai/nn/lr_scheduler.py (hunk 5 above)
Comment (stephenroller) [reply parlai/nn/lr_scheduler.py:504]: It makes it so the LR never hits exactly 0 (at most 1e-6 * the peak LR)
